### PR TITLE
Merges encoding to next

### PR DESCRIPTION
**Pull request by stevemk14ebr**
follow up to https://github.com/aquynh/capstone/pull/1187
https://github.com/aquynh/capstone/pull/520

### CODE REVIEW

**Comment by stevemk14ebr** [comment]
idk why the python binding fails, help needed

File: tests/test_x86.c [hunk 1]
```diff
@@ -152,8 +152,19 @@ static void print_insn_detail(csh ud, cs_mode mode, cs_insn *ins)
 
 	printf("\taddr_size: %u\n", x86->addr_size);
 	printf("\tmodrm: 0x%x\n", x86->modrm);
+	if (x86->encoding.modrm_offset != 0) {
+		printf("\tmodrm_offset: 0x%x\n", x86->encoding.modrm_offset);
+	}
+	
 	printf("\tdisp: 0x%x\n", x86->disp);
```
Comment (aquynh) [inline tests/test_x86.c]: this needs to be fixed, because this field is now int64

File: include/capstone/x86.h [hunk 2]
```diff
@@ -323,13 +336,12 @@ typedef struct cs_x86 {
 	// SIB value, or 0 when irrelevant.
 	uint8_t sib;
 
-	// Displacement value, or 0 when irrelevant.
-	int32_t disp;
+	// Displacement value, valid if encoding.disp_offset != 0
+	int64_t disp;
```
Comment (aquynh) [inline include/capstone/x86.h]: cstool_x86.c prints this field as int32, so needs to be fixed.

**Comment by aquynh** [comment]
```
idk why the python binding fails, help needed. C passes tests and my tests as well
```

can you confirm that bindings/python/test_x86.py runs without any issues?

**Comment by stevemk14ebr** [comment]
@aquynh i can do git stuff and im good at C, python i'm clueless at im fairly sure it won't work locally either. I just copied the changes made from master, which do work there. I'll keep working on it in my spare time but if a more skilled python dev like yourself could take a quick look to see what's wrong this would be solved fast.

**Comment by aquynh** [comment]
I see, let me look at that

**Comment by aquynh** [comment]
the Python binding is broken because it misses those fields.

a patch for master branch (currently broken) is like this:

```
diff --git a/bindings/python/capstone/__init__.py b/bindings/python/capstone/__init__.py
index de6e0c1d..3f0cd5fc 100644
--- a/bindings/python/capstone/__init__.py
+++ b/bindings/python/capstone/__init__.py
@@ -552,7 +552,9 @@ class CsInsn(object):
             (self.prefix, self.opcode, self.rex, self.addr_size, \
                 self.modrm, self.sib, self.disp, \
                 self.sib_index, self.sib_scale, self.sib_base, self.sse_cc, \
-                self.avx_cc, self.avx_sae, self.avx_rm, self.operands, self.encoding) = x86.get_arch_info(self._detail.arch.x86)
+                self.avx_cc, self.avx_sae, self.avx_rm, self.modrm_offset, \
+                self.disp_offset, self.disp_size, self.imm_offset, \
+                self.imm_size, self.operands) = x86.get_arch_info(self._detail.arch.x86)
         elif arch == CS_ARCH_MIPS:
                 self.operands = mips.get_arch_info(self._detail.arch.mips)
         elif arch == CS_ARCH_PPC:
diff --git a/bindings/python/capstone/x86.py b/bindings/python/capstone/x86.py
index 479060cc..298d3c64 100644
--- a/bindings/python/capstone/x86.py
+++ b/bindings/python/capstone/x86.py
@@ -81,6 +81,7 @@ def get_arch_info(a):
     return (a.prefix[:], a.opcode[:], a.rex, a.addr_size, \
             a.modrm, a.sib, a.disp, a.sib_index, a.sib_scale, \
             a.sib_base, a.sse_cc, a.avx_cc, a.avx_sae, a.avx_rm, \
-            copy_ctypes_list(a.operands[:a.op_count]), \
-            copy_ctypes(a.encoding))
+            a.encoding.modrm_offset, a.encoding.disp_offset, a.encoding.disp_size, \
+            a.encoding.imm_offset, a.encoding.imm_size, \
+            copy_ctypes_list(a.operands[:a.op_count]))
```

**Comment by aquynh** [comment]
i think you can just apply the above patch to fix Python binding of this next branch.

**Comment by stevemk14ebr** [comment]
fix is a no-go :/

**Comment by aquynh** [comment]
you missed the field `eflags`. here is a patch for Python binding:

```
diff --git a/bindings/python/capstone/__init__.py b/bindings/python/capstone/__init__.py
index 7dee05ad..eff46c12 100644
--- a/bindings/python/capstone/__init__.py
+++ b/bindings/python/capstone/__init__.py
@@ -631,9 +631,9 @@ class CsInsn(object):
             (self.prefix, self.opcode, self.rex, self.addr_size, \
                 self.modrm, self.sib, self.disp, \
                 self.sib_index, self.sib_scale, self.sib_base, self.xop_cc, self.sse_cc, \
-                self.avx_cc, self.avx_sae, self.avx_rm, self.modrm_offset, \
-                self.disp_offset, self.disp_size, self.imm_offset, \
-                self.imm_size, self.operands) = x86.get_arch_info(self._detail.arch.x86)
+                self.avx_cc, self.avx_sae, self.avx_rm, self.eflags, \
+                self.modrm_offset, self.disp_offset, self.disp_size, self.imm_offset, self.imm_size, \
+                self.operands) = x86.get_arch_info(self._raw.detail.contents.arch.x86)
         elif arch == CS_ARCH_M68K:
                 (self.operands, self.op_size) = m68k.get_arch_info(self._raw.detail.contents.arch.m68k)
         elif arch == CS_ARCH_MIPS:
diff --git a/bindings/python/capstone/x86.py b/bindings/python/capstone/x86.py
index 8c81cb79..ede5f856 100644
--- a/bindings/python/capstone/x86.py
+++ b/bindings/python/capstone/x86.py
@@ -1,7 +1,7 @@
 # Capstone Python bindings, by Nguyen Anh Quynnh <aquynh@gmail.com>

 import ctypes
-from . import copy_ctypes, copy_ctypes_list
+from . import copy_ctypes_list
 from .x86_const import *

 # define the API
@@ -43,6 +43,7 @@ class X86Op(ctypes.Structure):
     def mem(self):
         return self.value.mem

+
 class CsX86Encoding(ctypes.Structure):
     _fields_ = (
         ('modrm_offset', ctypes.c_uint8),
@@ -79,7 +80,6 @@ def get_arch_info(a):
     return (a.prefix[:], a.opcode[:], a.rex, a.addr_size, \
             a.modrm, a.sib, a.disp, a.sib_index, a.sib_scale, \
             a.sib_base, a.xop_cc, a.sse_cc, a.avx_cc, a.avx_sae, a.avx_rm, a.eflags, \
-            a.encoding.modrm_offset, a.encoding.disp_offset, a.encoding.disp_size, \
-            a.encoding.imm_offset, a.encoding.imm_size, \
+            a.encoding.modrm_offset, a.encoding.disp_offset, a.encoding.disp_size, a.encoding.imm_offset, a.encoding.imm_size, \
             copy_ctypes_list(a.operands[:a.op_count]))

```

**Comment by aquynh** [comment]
here is a patch for warnings regarding printing int64 fields.

```
diff --git a/cstool/cstool_x86.c b/cstool/cstool_x86.c
index ae551621..f5d2bab9 100644
--- a/cstool/cstool_x86.c
+++ b/cstool/cstool_x86.c
@@ -195,7 +195,7 @@ void print_insn_detail_x86(csh ud, cs_mode mode, cs_insn *ins)
        printf("\trex: 0x%x\n", x86->rex);
        printf("\taddr_size: %u\n", x86->addr_size);
        printf("\tmodrm: 0x%x\n", x86->modrm);
-       printf("\tdisp: 0x%x\n", x86->disp);
+       printf("\tdisp: 0x%" PRIx64 "\n", x86->disp);

        // SIB is not available in 16-bit mode
        if ((mode & CS_MODE_16) == 0) {
diff --git a/tests/test_x86.c b/tests/test_x86.c
index dadffb3e..a305551e 100644
--- a/tests/test_x86.c
+++ b/tests/test_x86.c
@@ -156,7 +156,7 @@ static void print_insn_detail(csh ud, cs_mode mode, cs_insn *ins)
                printf("\tmodrm_offset: 0x%x\n", x86->encoding.modrm_offset);
        }

-       printf("\tdisp: 0x%x\n", x86->disp);
+       printf("\tdisp: 0x%" PRIx64 "\n", x86->disp);
        if (x86->encoding.disp_offset != 0) {
                printf("\tdisp_offset: 0x%x\n", x86->encoding.disp_offset);
        }
```

**Comment by aquynh** [comment]
an extra patch to change the data type of disp field of Python binding:

```
diff --git a/bindings/python/capstone/x86.py b/bindings/python/capstone/x86.py
index 8c81cb79..548581c5 100644
--- a/bindings/python/capstone/x86.py
+++ b/bindings/python/capstone/x86.py
@@ -60,7 +60,7 @@ class CsX86(ctypes.Structure):
         ('addr_size', ctypes.c_uint8),
         ('modrm', ctypes.c_uint8),
         ('sib', ctypes.c_uint8),
-        ('disp', ctypes.c_int32),
+        ('disp', ctypes.c_int64),
         ('sib_index', ctypes.c_uint),
         ('sib_scale', ctypes.c_int8),
         ('sib_base', ctypes.c_uint),
```

**Comment by stevemk14ebr** [comment]
yay we made it, thank you

**Comment by aquynh** [comment]
did you push those fixes here yet?

**Comment by aquynh** [comment]
ok saw it

File: tests/test_x86.c [hunk 3]
```diff
@@ -152,8 +152,19 @@ static void print_insn_detail(csh ud, cs_mode mode, cs_insn *ins)
...
+		printf("\tmodrm_offset: 0x%x\n", x86->encoding.modrm_offset);
+	}
+	
+	printf("\tdisp: 0x%" PRIx64 "\n", x86->disp);
+	if (x86->encoding.disp_offset != 0) {
+		printf("\tdisp_offset: 0x%x\n", x86->encoding.disp_offset);
+	}
+	
+	if (x86->encoding.disp_size != 0) {
+	    printf("\tdisp_size: 0x%x\n", x86->encoding.disp_size);
```
Comment (aquynh) [inline tests/test_x86.c]: indentation with tab here, please

File: arch/X86/X86ATTInstPrinter.c [hunk 4]
```diff
@@ -738,7 +739,10 @@ static void printOperand(MCInst *MI, unsigned OpNo, SStream *O)
 				MI->flat_insn->detail->x86.operands[MI->flat_insn->detail->x86.op_count].imm = imm;
 
 				if (opsize > 0)
-					MI->flat_insn->detail->x86.operands[MI->flat_insn->detail->x86.op_count].size = (uint8_t)opsize;
+				{
+					MI->flat_insn->detail->x86.operands[MI->flat_insn->detail->x86.op_count].size = opsize;
```
Comment (aquynh) [inline arch/X86/X86ATTInstPrinter.c]: i think we still need to cast to `uint8_t` here.

**Comment by aquynh** [comment]
could you put your name & Martin's name into CREDITS.TXT for this change?

**Comment by stevemk14ebr** [comment]
done

File: arch/X86/X86Mapping.h [hunk 5]
```diff
@@ -69,6 +69,6 @@ void X86_reg_access(const cs_insn *insn,
 		cs_regs regs_write, uint8_t *regs_write_count);
 
 // given the instruction id, return the size of its immediate operand (or 0)
-int X86_immediate_size(unsigned int id);
+int X86_immediate_size(unsigned int id, int *enc_size);
```
Comment (aquynh) [inline arch/X86/X86Mapping.h]: please change the data type of enc_size to `uint8_t *`.

**Comment by stevemk14ebr** [comment]
i believe this is up to date and ready to go, are there any other changes needed?

**Comment by aquynh** [comment]
there are still some minor issues, but i can fix them.

merged, thanks!

**Comment by aquynh** [comment]
i made some fixes on coding style here: https://github.com/aquynh/capstone/commit/3fef42f345c63d64a49890b3f525ed2915fd3f11

**Comment by aquynh** [comment]
i reverted this encoding feature on "master", so only "next" branch has it.

@oleavr, can you confirm that the next branch now is good for Frida?

**Comment by oleavr** [comment]
@aquynh Just ported Frida to latest -next, and the good news is that all of our X86Relocator tests are green. The bad news is that it broke Stalker, so either there's an API change I'm not aware of, or there's a regression somewhere in Capstone. I'll let you know when I've identified the root cause.

**Pull request merged**